src/linalg/exp.rs

```diff
@@ -0,0 +1,479 @@
+//! This module provides the matrix exponent (exp) function to square matrices.
+//!
+use crate::{
+    base::{
+        allocator::Allocator,
+        dimension::{DimMin, DimMinimum, DimName},
+        DefaultAllocator,
+    },
+    try_convert, ComplexField, MatrixN, RealField,
+};
+
+// https://github.com/scipy/scipy/blob/c1372d8aa90a73d8a52f135529293ff4edb98fc8/scipy/sparse/linalg/matfuncs.py
+struct ExpmPadeHelper<N, R>
+where
+    N: RealField,
+    R: DimName + DimMin<R>,
+    DefaultAllocator: Allocator<N, R, R> + Allocator<(usize, usize), DimMinimum<R, R>>,
+{
+    use_exact_norm: bool,
+    ident: MatrixN<N, R>,
+
+    a: MatrixN<N, R>,
+    a2: Option<MatrixN<N, R>>,
+    a4: Option<MatrixN<N, R>>,
+    a6: Option<MatrixN<N, R>>,
+    a8: Option<MatrixN<N, R>>,
+    a10: Option<MatrixN<N, R>>,
+
+    d4_exact: Option<N>,
+    d6_exact: Option<N>,
+    d8_exact: Option<N>,
+    d10_exact: Option<N>,
+
+    d4_approx: Option<N>,
+    d6_approx: Option<N>,
+    d8_approx: Option<N>,
+    d10_approx: Option<N>,
+}
+
+impl<N, R> ExpmPadeHelper<N, R>
+where
+    N: RealField,
+    R: DimName + DimMin<R>,
+    DefaultAllocator: Allocator<N, R, R> + Allocator<(usize, usize), DimMinimum<R, R>>,
+{
+    fn new(a: MatrixN<N, R>, use_exact_norm: bool) -> Self {
+        ExpmPadeHelper {
+            use_exact_norm,
+            ident: MatrixN::<N, R>::identity(),
+            a,
+            a2: None,
+            a4: None,
+            a6: None,
+            a8: None,
+            a10: None,
+            d4_exact: None,
+            d6_exact: None,
+            d8_exact: None,
+            d10_exact: None,
+            d4_approx: None,
+            d6_approx: None,
+            d8_approx: None,
+            d10_approx: None,
+        }
+    }
+
+    fn a2(&self) -> &MatrixN<N, R> {
+        if self.a2.is_none() {
+            let ap = &self.a2 as *const Option<MatrixN<N, R>> as *mut Option<MatrixN<N, R>>;
```

Comment:
Casting a non-mutable reference to a to a mutable pointer is always UB. See https://doc.rust-lang.org/nomicon/transmutes.html.
You can either use `std::cell::Cell` or `std::cell::RefCell` instead of this unsafe cast.

What is the reason why you can't take `&mut self` as parameter here?

Reply:
Now fixed!